lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/BalanceSpec.hs

```diff
@@ -1939,60 +1999,136 @@ mkMinCoinValueFor = \case
+genMockComputeMinimumCost = arbitraryBoundedEnum
+
+shrinkMockComputeMinimumCost
+    :: MockComputeMinimumCost -> [MockComputeMinimumCost]
+shrinkMockComputeMinimumCost = \case
+    MockComputeMinimumCostZero ->
+        []
+    MockComputeMinimumCostLinear ->
+        [MockComputeMinimumCostZero]
+
+unMockComputeMinimumCost
+    :: MockComputeMinimumCost -> (SelectionSkeleton -> Coin)
+unMockComputeMinimumCost = \case
+    MockComputeMinimumCostZero ->
+        computeMinimumCostZero
+    MockComputeMinimumCostLinear ->
+        computeMinimumCostLinear
+
+computeMinimumCostZero :: SelectionSkeleton -> Coin
+computeMinimumCostZero = const $ Coin 0
+
+computeMinimumCostLinear :: SelectionSkeleton -> Coin
+computeMinimumCostLinear s
     = Coin
     $ fromIntegral
     $ skeletonInputCount s
     + F.length (skeletonOutputs s)
     + F.length (skeletonChange s)
 
-type MakeChangeData =
-    MakeChangeCriteria MinCoinValueFor MockTokenBundleSizeAssessor
+--------------------------------------------------------------------------------
+-- Computing selection limits
+--------------------------------------------------------------------------------
 
-data MockTokenBundleSizeAssessor
-    = NoBundleSizeLimit
+data MockComputeSelectionLimit
+    = MockComputeSelectionLimitNone
+    | MockComputeSelectionLimit Int
+    deriving (Eq, Show)
+
+genMockComputeSelectionLimit :: Gen MockComputeSelectionLimit
+genMockComputeSelectionLimit = oneof
+    [ pure MockComputeSelectionLimitNone
+    , MockComputeSelectionLimit <$> sized (\n -> choose (1, max 1 n))
+    ]
+
+shrinkMockComputeSelectionLimit
+    :: MockComputeSelectionLimit -> [MockComputeSelectionLimit]
+shrinkMockComputeSelectionLimit = \case
+    MockComputeSelectionLimitNone ->
+        []
+    MockComputeSelectionLimit n ->
+        MockComputeSelectionLimit <$> filter (> 0) (shrink n)
+
+unMockComputeSelectionLimit
+    :: MockComputeSelectionLimit -> ([TxOut] -> SelectionLimit)
+unMockComputeSelectionLimit = \case
+    MockComputeSelectionLimitNone ->
+        const NoLimit
+    MockComputeSelectionLimit n ->
+        const $ MaximumInputLimit n
+
+--------------------------------------------------------------------------------
+-- Assessing token bundle sizes
+--------------------------------------------------------------------------------
+
+data MockAssessTokenBundleSize
+    = MockAssessTokenBundleSizeUnlimited
       -- ^ Indicates that there is no limit on a token bundle's size.
-    | BundleAssetCountUpperLimit Int
+    | MockAssessTokenBundleSizeUpperLimit Int
       -- ^ Indicates an inclusive upper bound on the number of assets in a
       -- token bundle.
     deriving (Eq, Show)
 
-mkBundleSizeAssessor
-    :: MockTokenBundleSizeAssessor -> TokenBundleSizeAssessor
-mkBundleSizeAssessor m = TokenBundleSizeAssessor $ case m of
-    NoBundleSizeLimit ->
+genMockAssessTokenBundleSize :: Gen MockAssessTokenBundleSize
+genMockAssessTokenBundleSize =
+    -- TODO:
```

Comment:
Is there a reason why we don't?

Reply:
> Is there a reason why we don't?

If I recall correctly, at the time we wrote `prop_performSelection` we believed it would complicate the property test too much to impose a token bundle size limit, so we always generated criteria without them.

I'm actually planning to improve this generator in the next PR: https://github.com/input-output-hk/cardano-wallet/pull/2909